procedures/unit-testing-basics.ipf

```diff
@@ -1311,171 +1730,237 @@ Function RunTest(procWinList, [name, testCase, enableJU, enableTAP, enableRegExp
+			getLocalHooks(s.procHooks, s.procWin)
 
-		NVAR/SDFR=dfr error_count
+			s.juProps.testCaseList = s.testCaseList
+			s.juProps.testSuiteNumber = s.i
+			ExecuteHooks(TEST_SUITE_BEGIN_CONST, s.procHooks, s.juProps, s.procWin, s.procWin)
+
+			NVAR/SDFR=dfr error_count
 
-		numItemsFFN = ItemsInList(testCaseList)
-		for(j = numItemsFFN-1; j >= 0; j -= 1)
-			fullFuncName = StringFromList(j, testCaseList)
-			FUNCREF TEST_CASE_PROTO TestCaseFunc = $fullFuncName
+			s.numItemsFFN = ItemsInList(s.testCaseList)
+		else
+			s.numItemsFFN = 1
+		endif
 
-			// get Description and Directive of current Function for TAP
-			tap_skipCase = 0
-			if(TAP_IsOutputEnabled())
-				tap_skipCase = TAP_GetNotes(fullFuncName)
+		for(j = s.numItemsFFN - 1; j >= 0; j -= 1)
+			s.j = j
+
+			if(!reentry)
+				s.fullFuncName = StringFromList(s.j, s.testCaseList)
+				FUNCREF TEST_CASE_PROTO TestCaseFunc = $s.fullFuncName
+
+				// get Description and Directive of current Function for TAP
+				s.tap_skipCase = 0
+				if(TAP_IsOutputEnabled())
+					s.tap_skipCase = TAP_GetNotes(s.fullFuncName)
+				endif
 			endif
 
-			if(!tap_skipCase)
-				ExecuteHooks(TEST_CASE_BEGIN_CONST, procHooks, juProps, fullFuncName, procWin)
+			if(!s.tap_skipCase || reentry)
+
+				if(!reentry)
+					ExecuteHooks(TEST_CASE_BEGIN_CONST, s.procHooks, s.juProps, s.fullFuncName, s.procWin)
+				else
 
+					DFREF dfSave = $PKG_FOLDER_SAVE
+					RestoreState(dfSave, s)
+					// restore all loop counters
+					i = s.i
+					j = s.j
+					DFREF dfSave = $""
+					ClearReentrytoUTF()
+					reentry = 0
+
+					// set reentry function
+					SVAR reentryFuncName = dfr:BCKG_ReentryFunc
+					FUNCREF TEST_CASE_PROTO TestCaseFunc = $reentryFuncName
+
+				endif
 				try
 					ClearRTError()
 					TestCaseFunc(); AbortOnRTE
 				catch
 					// only complain here if the error counter if the abort happened not in our code
 					if(!shouldDoAbort())
 						message = GetRTErrMessage()
-						err = GetRTError(1)
-						EvaluateRTE(err, message, V_AbortCode, fullFuncName, TEST_CASE_TYPE, procWin)
+						s.err = GetRTError(1)
+						EvaluateRTE(s.err, message, V_AbortCode, s.fullFuncName, TEST_CASE_TYPE, s.procWin)
 						incrError()
 					endif
 
 					if(shouldDoAbort())
 						// abort condition is on hold while in catch/endtry, so all cleanup must happen here
-						ExecuteHooks(TEST_CASE_END_CONST, procHooks, juProps, fullFuncName, procWin, param = keepDataFolder)
+						ExecuteHooks(TEST_CASE_END_CONST, s.procHooks, s.juProps, s.fullFuncName, s.procWin, param = s.keepDataFolder)
 
-						ExecuteHooks(TEST_SUITE_END_CONST, procHooks, juProps, procWin, procWin)
+						ExecuteHooks(TEST_SUITE_END_CONST, s.procHooks, s.juProps, s.procWin, s.procWin)
 
-						ExecuteHooks(TEST_END_CONST, hooks, juProps, name, NO_SOURCE_PROCEDURE, param = allowDebug)
+						ExecuteHooks(TEST_END_CONST, s.hooks, s.juProps, s.name, NO_SOURCE_PROCEDURE, param = s.allowDebug)
```

Comment:
@MichaelHuth Do we need to do a `QuitOnAutoRunFull` here?

Reply:
Yes, I missed that return.